File: lang/ruby/src/fun/vac/util/Sequences.rb
require "fun/vac/algorithms/sorting/InsertionSort"
require "fun/vac/util/RandomFactory"

module Sequences

    def self.inspect(arr)
        puts("[")
        arr.each_index do |i|
            print("\t#%04X  -->  %d\n" % [i, arr[i]])
        end
        puts("]")
    end

    def self.isSorted(arr)
        for i in 1 ... arr.length
            if arr[i] < arr[i - 1]
                return false
            end
        end

        return true
    end

    def self.parityChecksum(arr)
        checksum = 0

        for v in arr
            checksum ^= v
        end

        return checksum
    end

    def self.reverse(arr)
        k = 0

        length = arr.length
        for i in 0 ... arr.length >> 1
            k = length - i - 1
            arr[i], arr[k] = arr[k], arr[i]
        end
    end

    def self.shuffle(arr)
        k = 0
        length = arr.length

        RandomFactory.seed()
        for i in 0 ... arr.length
            k = RandomFactory.integerN(i, length)
            arr[i], arr[k] = arr[k], arr[i]
        end
    end

    def self.sort(arr)
        InsertionSort.sort(arr)
    end
end

File: lang/php/src/fun/vac/util/Sequences.php
<?php
    namespace fun\vac\util\Sequences;

    require_once "fun/vac/algorithms/sorting/InsertionSort.php";
    require_once "fun/vac/util/RandomFactory.php";

    use fun\vac\algorithms\sorting\InsertionSort;
    use fun\vac\util\RandomFactory;

    function inspect(array &$arr): void {
        print("[\n");
        foreach ($arr as $i => $v) {
            printf("\t#%04X  -->  %d\n", $i, $v);
        }
        print("]\n");
    }

    function isSorted(array &$arr): bool {
        for ($i = 1, $length = count($arr); $i < $length; $i++) {
            if ($arr[$i] < $arr[$i - 1]) {
                return false;
            }
        }

        return true;
    }

    function parityChecksum(array &$arr): int {
        $checksum = 0;

        foreach ($arr as $v) {
            $checksum ^= $v;
        }

        return $checksum;
    }

    function reverse(array &$arr): void {
        $k;
        $temp;

        for ($i = 0, $bound = count($arr) >> 1, $length = count($arr); $i < $bound; $i++) {
            $k = $length - $i - 1;
            $temp = $arr[$i];
            $arr[$i] = $arr[$k];
            $arr[$k] = $temp;
        }
    }

    function shuffle(array &$arr): void {
        $k;
        $temp;

        RandomFactory\seed();
        for ($i = 0, $length = count($arr); $i < $length; $i++) {
            $k = RandomFactory\integerN($i, $length);
            $temp = $arr[$i];
            $arr[$i] = $k;
            $k = $temp;
        }
    }

    function sort(array &$arr): void {
        InsertionSort\sort($arr);
    }
?>
